File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.py
# recursive binary search to find min
# using a helper function, search, to do the binary search
# use low and high pointers to keep track of current subarray end positions
# keep dividing until 1 number is left (i == j)
# if mid < high (right half is sorted), then move to left half (including mid)
# else if mid >= high (left half is sorted), then move to right half (excluding mid)

# by always discarding the sorted half, we eventually reach the rotation point 

# include/exclude mid when discarding left/right halfs to avoid infinite cycles
# when 2 numbers are left, can't properly discard left half because mid is rounded down
# so we must do mid+1 when setting it as the new low

class Solution:
    def findMin(self, nums: List[int]) -> int:
        def search(nums, low, high):
            if low == high:
                return nums[low]

            mid = math.floor((low + high) / 2)
            if (nums[mid] <= nums[high]):
                return search(nums, low, mid)
            else:
                return search(nums, mid+1, high)
            
        return search(nums, 0, len(nums)-1)

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */

// binary search but we look for the rotation point (highest number before the lowest number)
// the min and highest values are at the rotation point
// go until low and high reach the same point (the min value)
// when mid > high (right half is rotated), then search the right half
// otherwise left half contains the rotation, so search the left half

// but keep mid when searching the left half in case that mid is the lowest point
// because we use Math.floor to round down mid, mid could end up being the min
// so we don't want to accidently discard it when going down into the left half
// or go to a point where there are no numbers left (e.g think about the [1, 2] case)

// return nums[high] or nums[low] since that'll be the min point

var findMin = function(nums) {
    let low = 0;
    let high = nums.length - 1;

    while (low < high) {
        const mid = Math.floor((low + high) / 2);
        if (nums[mid] > nums[high]) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }

    return nums[high];
};